### ISSUE TITLE
[BUG] VSCode freezes on file change in large repo

### ISSUE THREAD

**Issue by christianvuerings** [jest-community/vscode-jest#1196]
**Describe the bug**

VSCode freezes when making a file change (create / rename / delete) in a repo with lots of tests.

**To Reproduce**
1. Clone repo with many test files: `git clone https://github.com/christianvuerings/vscode-jest-large-repo.git`
2. Open VSCode with that repo: `cd vscode-jest-large-repo; code .`
3. Run all tests in test explorer (running just 1 test also repros the issue)
![Image](https://github.com/user-attachments/assets/73dfb52e-cff9-449e-9f02-26486e5ebe08)
4. After tests finish to run, rename a non-test file `src/main.js` to `src/main_renamed.js`: `mv src/main.js src/main_renamed.js`

**Expected behavior**
VSCode doesn't freeze

**Actual behavior**
VSCode freezes for 5-10 seconds

**Screenshots**

First part: shows freezing issue with 2 separate renames.
Second part: renaming works smoothly without the VSCode Jest extension

https://github.com/user-attachments/assets/12033b73-72e2-4a4f-9754-7ac703b23b2c

**Environment:**
- vscode-jest version: `v6.4.0`
- `node -v`: `v18.16.1`
- `npm -v` or `yarn --version`: `9.5.1`
- jest or react-scripts (if you haven’t ejected) version: `Jest 29.7.0`
- your vscode-jest settings:
    - jest.jestCommandLine? `None`
    - jest.runMode? `on-demand`
    - jest.outputConfig? `None`
    - anything else that you think might be relevant? Set `deferred` to `true`
- Operating system: `Mac OS 14.6.1`

**Prerequisite**
- are you able to run jest from the command line? `Yes`
- where do you run jest CLI from? `Root directory of the project`
- how do you run your tests from the command line? `npm run test`

**Additional context**

When debugging the root cause seems to be `updateTestFileList`:
https://github.com/jest-community/vscode-jest/blob/fae16e86446379f907079dab9129041869d1c585/src/JestExt/core.ts#L744-L765

Which gets called on every file create, delete or rename
https://github.com/jest-community/vscode-jest/blob/fae16e86446379f907079dab9129041869d1c585/src/JestExt/core.ts#L767-L775

If we comment out the contents of `updateTestFileList` everything works smoothly

**Comment by connectdotz** [comment]
@christianvuerings, this is an interesting use case... The behavior you see in `JestExt/core.ts` is intended to keep the internal test file list up-to-date upon file lifecycle changes. But I can see this could impact a large repo even with the "on-demand" setting. 

I guess we could explore a "just-in-time" updateTestFileList(), which only executes, if needed, when the next test run is scheduled...

**Comment by christianvuerings** [comment]
@connectdotz Were you thinking something along these lines? https://github.com/jest-community/vscode-jest/pull/1233

**Comment by connectdotz** [comment]
@christianvuerings, yes indeed. But I am still struggling with a few questions, let's discuss in your PR. Thanks for your quick action!!👍
